ocaml/xapi/tar_ext.ml

```diff
@@ -0,0 +1,128 @@
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+open Helpers
+
+module D = Debug.Make (struct let name = __MODULE__ end)
+
+open D
+
+let dir_perm = 0o755
+
+let dir_size = 4096L
+
+let inode_size = 4096L
+
+let ( + ) = Int64.add
+
+let ( // ) = Filename.concat
+
+exception Illegal_file_path of string
+
+exception Unsupported_file_type of string
+
+exception File_exceeds_max_size_limit of string
+
+exception File_incomplete of string
+
+exception File_corrupted of string
+
+exception File_parsed_failed of string
+
+let unpack dir acc_size ifd head =
+  let assert_file_path_is_legal file_path =
+    if
+      List.exists
+        (fun s ->
+          String.equal Filename.current_dir_name s
+          || String.equal Filename.parent_dir_name s
+        )
+        (String.split_on_char '/' file_path)
+    then
+      raise (Illegal_file_path file_path)
+  in
+  let assert_file_not_exceed_max_size acc_size =
+    if Int64.compare acc_size !Xapi_globs.bundle_file_max_size > 0 then
+      raise
+        (File_exceeds_max_size_limit
+           (Printf.sprintf
+              "Current total unpacked file size: %Ld, exceeds max size limit: \
+               %Ld"
+              acc_size
+              !Xapi_globs.bundle_file_max_size
+           )
+        )
+  in
+  let file_name = head.Tar.Header.file_name in
+  debug "Unpacking tar file: %s" file_name ;
+  assert_file_path_is_legal file_name ;
+  let file_path = dir // file_name in
+  match head.Tar.Header.link_indicator with
+  | Tar.Header.Link.Normal ->
+      let file_size = head.Tar.Header.file_size in
+      let new_acc_size = acc_size + file_size + inode_size in
+      Unixext.with_file file_path [Unix.O_WRONLY; Unix.O_CREAT; Unix.O_TRUNC]
+        head.Tar.Header.file_mode (fun ofd ->
+          assert_file_not_exceed_max_size new_acc_size ;
+          let real_copied_size = Unixext.copy_file ~limit:file_size ifd ofd in
+          if real_copied_size <> file_size then
+            raise (File_incomplete "Copied size is insistent with file size")
+      ) ;
+      new_acc_size
+  | Tar.Header.Link.Directory ->
+      let new_acc_size = acc_size + dir_size + inode_size in
+      assert_file_not_exceed_max_size new_acc_size ;
+      Unixext.mkdir_rec file_path head.Tar.Header.file_mode ;
+      new_acc_size
+  | Hard
+  | Symbolic
+  | Character
+  | Block
+  | FIFO
+  | GlobalExtendedHeader
+  | PerFileExtendedHeader
+  | LongLink ->
+      raise
+        (Unsupported_file_type
+           (Tar.Header.Link.to_string head.Tar.Header.link_indicator)
+        )
+
+let unpack_tar_file ~dir ~ifd =
```

Comment:
I think it is worth using `Result` instead of `exceptions` as the return of this function. And those exceptions from the libs can be converted into a single type with other errors.
Some necessary info needs to be kept/logged for debugging purpose.